src/runtime-rs/crates/hypervisor/src/device/tap.rs

```diff
@@ -0,0 +1,266 @@
+    }
+}
+fn create_socket() -> Result<UdpSocket> {
+    // This is safe since we check the return value.
+    let sock = unsafe { libc::socket(libc::AF_INET, libc::SOCK_DGRAM, 0) };
+    if sock < 0 {
+        return Err(Error::CreateSocket(IoError::last_os_error()));
+    }
+
+    // This is safe; nothing else will use or hold onto the raw sock fd.
+    Ok(unsafe { UdpSocket::from_raw_fd(sock) })
+}
+
+// Returns a byte vector representing the contents of a null terminated C string which
+// contains if_name.
+pub fn build_terminated_if_name(if_name: &str) -> Result<[i8; IFACE_NAME_MAX_LEN]> {
+    // Convert the string slice to bytes, and shadow the variable,
+    // since we no longer need the &str version.
+    let if_name_bytes = if_name.as_bytes();
+
+    if if_name_bytes.len() >= IFACE_NAME_MAX_LEN {
+        return Err(Error::InvalidIfname);
+    }
+
+    let mut terminated_if_name = [0i8; IFACE_NAME_MAX_LEN];
+    for (i, &byte) in if_name_bytes.iter().enumerate() {
+        terminated_if_name[i] = byte as i8;
+    }
+
+    // 0 is the null terminator for i8 type
+    terminated_if_name[if_name_bytes.len()] = 0;
+
+    Ok(terminated_if_name)
+}
+
+impl Tap {
+    /// Create a TUN/TAP device given the interface name.
+    /// # Arguments
+    ///
+    /// * `if_name` - the name of the interface.
+    pub fn open_named(if_name: &str, multi_vq: bool) -> Result<Tap> {
+        let terminated_if_name = build_terminated_if_name(if_name)?;
+
+        //Initialize an `ifreq` structure with the given interface name
+        //and configure its flags for setting up a network interface.
+        let mut ifr: ifreq = unsafe { std::mem::zeroed() };
+        ifr.ifr_name = terminated_if_name;
+        ifr.ifr_ifru.ifru_flags =
+            (IFF_TAP | IFF_NO_PI | IFF_VNET_HDR | if multi_vq { IFF_MULTI_QUEUE } else { 0 })
+                as c_short;
+
+        Tap::create_tap_with_ifreq(&mut ifr)
+    }
+
+    fn create_tap_with_ifreq(ifr: &mut ifreq) -> Result<Tap> {
+        let fd = unsafe {
+            // Open calls are safe because we give a constant null-terminated
+            // string and verify the result.
+            libc::open(
+                b"/dev/net/tun\0".as_ptr() as *const c_char,
+                libc::O_RDWR | libc::O_NONBLOCK | libc::O_CLOEXEC,
+            )
+        };
+        if fd < 0 {
+            return Err(Error::OpenTun(IoError::last_os_error()));
+        }
+
+        // We just checked that the fd is valid.
+        let tuntap = unsafe { File::from_raw_fd(fd) };
+
+        // ioctl is safe since we call it with a valid tap fd and check the return
+        // value.
+        let ret = unsafe { ioctl_with_mut_ref(&tuntap, TUNSETIFF(), ifr) };
+
+        if ret < 0 {
+            return Err(Error::CreateTap(IoError::last_os_error()));
+        }
+        Ok(Tap {
+            tap_file: tuntap,
+            if_name: ifr.ifr_name,
+            if_flags: unsafe { ifr.ifr_ifru.ifru_flags },
```

Comment:
Why do we need an `unsafe` to read a field?

Reply:
We use unsafe because interacting with the C API, particularly when reading if_flags, involves handling a C union, which is considered unsafe in Rust.
> Why do we need an `unsafe` to read a field?